### PR TITLE
Modified metric.py, test_metric.py

### PR DESCRIPTION
**Pull request by z123z123d**
**Description:**
Modified metric.py, test_metric.py

### CODE REVIEW

File: tests/dataloader/test_metric.py [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+
+
+class FakeDataLoader:
+	def __init__(self):
+		self.vocab_list = ['<pad>', '<unk>', '<go>', '<eos>', 'what', 'how', 'here', 'do']
+		self.vocab_to_index = {'<pad>': 0, '<unk>': 1, '<go>': 2, '<eos>': 3, 'what': 4, 'how': 5, 'here': 6, 'do': 7}
+		self.vocab_size = 8
+
+	def trim_before_target(self, lists, target):
+		try:
```
Comment (hzhwcmhf) [inline tests/dataloader/test_metric.py]: use pytest.raises

File: tests/dataloader/test_metric.py [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+		dataloader = FakeDataLoader()
+		data = dataloader.get_data('reference_key', 'reference_len_key', 'gen_prob_key',\
+								   'gen_key', 'post_key', 'resp_key', 'res_key', \
+								   True, False)
+
+		data['reference_key'] = np.delete(data['reference_key'], 1)
+
+		pm = PerlplexityMetric(dataloader, 'reference_key', 'reference_len_key', 'gen_prob_key', full_check=True)
+
+		try:
```
Comment (hzhwcmhf) [inline tests/dataloader/test_metric.py]: Use pytest.raises

File: tests/dataloader/test_metric.py [hunk 3]
```diff
@@ -0,0 +1,585 @@
...
+		mc.add_metric(bcm)
+		mc.forward(data)
+		res = mc.close()
+
+		assert np.isclose(res['perplexity'], perplexity)
+		assert np.isclose(res['bleu'], bleu)
+
+if __name__ == '__main__':
+	T = TestMultiTurnDialogRecorder()
+	T.test_close()
```
Comment (hzhwcmhf) [inline tests/dataloader/test_metric.py]: add a new line at the end of file

File: tests/dataloader/test_metric.py [hunk 4]
```diff
@@ -0,0 +1,585 @@
...
+	def get_bleu(self, dataloader, input):
+		bleu_value = 0
+		sentence_num = 0
+		for gen_sen, resp_sen in zip(input['reference_key'], input['gen_key']):
+			gen_sen_processed = dataloader.trim_index(gen_sen)
+			resp_sen_processed = dataloader.trim_index(resp_sen)
+			bleu_value += sentence_bleu([resp_sen_processed[1 : ]], gen_sen_processed, \
+										smoothing_function=SmoothingFunction().method7)
+			sentence_num += 1
+		return bleu_value / sentence_num
```
Comment (hzhwcmhf) [inline tests/dataloader/test_metric.py]: corpus_blue is not always equals to the average of sentence_bleu. (Although it is right when there is only one sentence.) Refer to: http://www.nltk.org/_modules/nltk/translate/bleu_score.html
I think it may not necessary to check the bleu value of a random string, because you must implement the ground truth in the same way as the lib. I think that you may find some examples (include some corner cases), and give a predefined answer. For example, when gen == resp, the bleu_value must be 1

Comment (z123z123d) [reply tests/dataloader/test_metric.py]: But if 'metric.py' uses corpus_bleu function from ntlk, I think whatever corner cases won't make it fail. Or if it fails, the question will be in ntlk.

File: contk/metric/metric.py [hunk 5]
```diff
@@ -78,8 +78,10 @@ def forward(self, data):
 				if not np.allclose(expsum, [1] * single_length):
 					raise ValueError("data[gen_prob_key] must be processed after log_softmax.")
 
-			self.word_loss += -np.sum(gen_prob[i][\
-				list(range(single_length-1)), resp[i][1:single_length]])
+			# self.word_loss += -np.sum(gen_prob[i][\
+			# 	list(range(single_length-1)), resp[i][1:single_length]])
+			for x, y in zip(list(range(single_length - 1)), resp[i][1:single_length]):
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: This is a bottkeneck of performance. Try converting numpy.array to see which way will be faster. 
(By the way, you can use enumerate instead of zip(range(XX), XX))

File: contk/metric/metric.py [hunk 6]
```diff
@@ -195,7 +197,7 @@ def close(self):
 
 			* **bleu**: bleu value.
 		'''
-		return corpus_bleu(self.refs, self.hyps, smoothing_function=SmoothingFunction.method7)
+		return {"bleu": corpus_bleu(self.refs, self.hyps, smoothing_function=SmoothingFunction.method7)}
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: Have you test this line?
I think only ``SmoothingFunction().method7`` works

Comment (z123z123d) [reply contk/metric/metric.py]: Yes, `SmoothingFunction().method7` always works when I am testing.  But I will check `corpus_bleu` again.

**Comment by hzhwcmhf** [comment]
[CodeFactor](https://www.codefactor.io/repository/github/thu-coai/contk) found multiple issues last seen at fe56a918f51469cb77732c8474509958e6a5957d:

#### Variable name "y" doesn't conform to snake_case naming style
[contk\metric\metric.py:83
](https://github.com/thu-coai/contk/blob/fe56a918f51469cb77732c8474509958e6a5957d/contk/metric/metric.py#L83)
#### Consider using enumerate instead of iterating with range and len
[contk\metric\metric.py:300
](https://github.com/thu-coai/contk/blob/fe56a918f51469cb77732c8474509958e6a5957d/contk/metric/metric.py#L300)[contk\metric\metric.py:397
](https://github.com/thu-coai/contk/blob/fe56a918f51469cb77732c8474509958e6a5957d/contk/metric/metric.py#L397)
#### Variable name "x" doesn't conform to snake_case naming style
[contk\metric\metric.py:83
](https://github.com/thu-coai/contk/blob/fe56a918f51469cb77732c8474509958e6a5957d/contk/metric/metric.py#L83)

File: tests/dataloader/test_metric.py [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+		data = dataloader.get_data('reference_key', 'reference_len_key', 'gen_prob_key',\
+								   'gen_key', 'post_key', 'resp_key', 'res_key', \
+								   True, False)
+		ans = self.get_bleu(dataloader, data)
+
+		bcm = BleuCorpusMetric(dataloader, 'reference_key', 'res_key')
+		bcm.forward(data)
+
+		print (bcm.close()['bleu'])
+		print (ans)
```
Comment (hzhwcmhf) [inline tests/dataloader/test_metric.py]: print?

File: tests/wordvector/test_wordvector.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+from contk.dataloader import LanguageGeneration, MSCOCO
+from contk.metric import MetricBase
+from contk.wordvector.wordvector import WordVector
+from contk.wordvector.gloves import Glove
+import logging
+
+class TestWordVector():
+	def base_test_init(self, dl):
+		assert isinstance(dl, WordVector)
+		try:
```
Comment (hzhwcmhf) [inline tests/wordvector/test_wordvector.py]: pytest.raises

File: contk/metric/metric.py [hunk 9]
```diff
@@ -293,11 +297,11 @@ def forward(self, data):
 		post = data[self.post_key]
 		resp = data[self.resp_key]
 		gen = data[self.gen_key]
-		if post.shape[0] != resp.shape[0] or resp.shape[0] != gen.shape[0]:
+		if len(post) != len(resp) or len(resp) != len(gen):
 			raise ValueError("Batch num is not matched.")
-		for i in range(post.shape[0]):
-			self.post_list.append(self.dataloader.index_to_sen(post[i, 1:]))
-			self.resp_list.append(self.dataloader.index_to_sen(resp[i, 1:]))
+		for _, i in enumerate(range(len(post))):
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: ```for i, _ in enumerate(post):```

File: contk/metric/metric.py [hunk 10]
```diff
@@ -293,11 +297,11 @@ def forward(self, data):
 		post = data[self.post_key]
 		resp = data[self.resp_key]
 		gen = data[self.gen_key]
-		if post.shape[0] != resp.shape[0] or resp.shape[0] != gen.shape[0]:
+		if len(post) != len(resp) or len(resp) != len(gen):
 			raise ValueError("Batch num is not matched.")
-		for i in range(post.shape[0]):
-			self.post_list.append(self.dataloader.index_to_sen(post[i, 1:]))
-			self.resp_list.append(self.dataloader.index_to_sen(resp[i, 1:]))
+		for i, _ in enumerate(post):
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: ```_``` means unused variable, if you want to use it, give it a good name that others can understand it.

File: contk/metric/metric.py [hunk 11]
```diff
@@ -392,8 +402,8 @@ def forward(self, data):
 				Size: `[batch_size, gen_sentence_length]`.
 		'''
 		gen = data[self.gen_key]
-		for i in range(gen.shape[0]):
-			self.gen_list.append(self.dataloader.index_to_sen(gen[i]))
+		for i, sen in enumerate(gen):
```
Comment (hzhwcmhf) [inline contk/metric/metric.py]: ```for sen in gen``` because ```i``` is unused.
